### ISSUE TITLE
[iOS] Syncing new (active) vocabulary not working

### ISSUE THREAD

**Issue by Kwintenvdb** [ulangi/ulangi#90]
[Ulangi version 3.20.4 on iOS]

Synchronizing new active vocabular terms seems broken on the new iOS version via the Google Sheets add-on.

Setting existing vocabulary to **DELETED** in Google Sheets and pushing the changes does synchronize properly - the respective terms get deleted on the iOS app. However, setting that same vocabular back to **ACTIVE** and pushing the changes does not synchronize properly to the iOS app anymore - they remain deleted or are not downloaded at all. 

The "Syncing" notification is shown persistently. I assume there is some error happening in the background that is not handled.

Note: I have a rather large set of data that I am trying to sync. Perhaps the synchronization struggles with large amounts of data.

**Comment by minhloi** [comment]
Yes. I believe that's one of the limitations of Google Sheets add-on. Just make sure to sync less than 500 rows at once (I don't know the exact number) but the more change you accumulate the more likely it will fail.

**Comment by Kwintenvdb** [comment]
What is the solution for this now that nothing will sync anymore at all? Do I need to start fresh with an entirely new sheet? Or is there some way this could be solved on the server side so changes are batched properly?

**Comment by alexalxs** [comment]
Interested in this

**Comment by minhloi** [comment]
You can try creating another document and then insert new terms from there.

**Comment by Kwintenvdb** [comment]
It still seems to break down at a certain number of terms (currently 3760). Even if I sync them in batches of 100 each time, at around this amount of terms, the syncing simply remains stuck forever. There appears to be no solution other than completely starting fresh from an entirely new account. Even reinstalling the app will just have the syncing get stuck after downloading a certain amount of terms.

**Comment by minhloi** [comment]
This issue should not be related to the mobile app. If the app keeps on syncing, it might be because you're having too many deleted terms? Currently, deleted terms are not flushed periodically.

Nevertheless, syncing on the app should not stop working because it is a progressive synchronizer. It syncs only new changes so if you add 100 new terms, it will download only those 100 new terms to your device.

**Comment by minhloi** [comment]
Also there are daily quotas when using Google Sheets: https://developers.google.com/apps-script/guides/services/quotas

The add-on uses **URL Fetch calls** a lot so make sure you don't reach this quota.

**Comment by Kwintenvdb** [comment]
> This issue should not be related to the mobile app. ~If the app keeps on syncing, it might be because you're having too many deleted terms? Currently, deleted terms are not flushed periodically~
> 
> Did you mean syncing on Google Sheets stuck? There are 2 synchronizers: One is in the Google Sheets add-on, the other is in the app. If the syncing on Google Sheets stuck, it is the problem of the Google Sheets add-on.
> 
> Nevertheless, syncing on the app should not stop working because it is a progressive synchronizer. It syncs only new changes so if you add 100 new terms, it will download only those 100 new terms to your device. Meaning it does not download everything again. The app is able to handle unlimited amount of data.

My issue specifically relates to the mobile app. I have a Google Sheets spreadsheet with about 4000 terms, all of which I successfully pushed (incrementally in batches of 100 as to not overload the system).

However, when I try to download these 4000 terms in the iOS app, it hangs at around 3700 terms. And it does that every time, even after a fresh reinstall, it will hang at somewhere around this amount. When I tried it for the third time again yesterday, it remained stuck here (and still is):

![image](https://user-images.githubusercontent.com/6693073/96552999-cf85d580-12b4-11eb-8ef6-62fd396021d8.png)

The Google Sheets add-on appears to function completely fine still, I can edit and push new data without issue.

**Comment by minhloi** [comment]
I have to check your data to know what's wrong. Maybe you uploaded malformed data. But it's highly unlikely.  I used to sync more than 5000 terms with another user account and it worked fine.

Can you send me your account email with the password to this email: bug@ulangi.com. You can change your password to something else first.

**Comment by Kwintenvdb** [comment]
> I have to check your data to know what's wrong. Maybe you uploaded malformed data. But it's highly unlikely. I used to sync more than 5000 terms with another user account and it worked fine.
> 
> Can you send me your account email with the password to this email: [bug@ulangi.com](mailto:bug@ulangi.com). You can change your password to something else first.

I've sent you an email. I hope you can find the issue. Thanks for your efforts!

**Comment by minhloi** [comment]
@Kwintenvdb I found the problem.

When you insert **vocabularyText** with empty term like below, it will fail.  
```

[note: Magdalena ist schon zweimal am Knie operiert worden.]
```
---

It fails at this rule: **vocabularyTerm** is allowed to be a string, null, undefined but it is not allowed to empty (by default).
https://github.com/ulangi/ulangi/blob/2c21c09cfc1444098a6a206b9168b4cd94d3874a/packages/ulangi-local-database/src/preparers/VocabularyLocalDataRowPreparer.ts#L25-L27

When I add ```allow('')```, it now syncs everything.

Normally, the server does all the checks (filter out invalid cards) so that the mobile app will not download them; however, vocabularyTerm is only stored and used locally thus the server does not have this check.

Right now you can either wait for the fix or you can just fix your cards first.

**Comment by Kwintenvdb** [comment]
Thanks a bunch for finding the issue! I will fix my cards, no idea how the data error snuck in there.

**Comment by minhloi** [comment]
fixed in a4229a395a7b30f4bf8a0c5d47faca5b261a4442
